packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFInstanceManager.m

```diff
@@ -3,70 +3,143 @@
+  self = [self init];
+  if (self) {
+    _identifier = identifier;
+    _callback = callback;
+  }
+  return self;
+}
+
++ (void)attachToInstance:(NSObject *)instance
+          withIdentifier:(long)identifier
+                callback:(FWFOnDeallocCallback)callback {
+  FWFFinalizer *finalizer = [[FWFFinalizer alloc] initWithIdentifier:identifier callback:callback];
+  objc_setAssociatedObject(instance, _cmd, finalizer, OBJC_ASSOCIATION_RETAIN);
+}
+
++ (void)detachFromInstance:(NSObject *)instance {
+  objc_setAssociatedObject(instance, @selector(attachToInstance:withIdentifier:callback:), nil,
+                           OBJC_ASSOCIATION_ASSIGN);
+}
+
+- (void)dealloc {
+  self.callback(self.identifier);
+}
+@end
 
 @interface FWFInstanceManager ()
 @property dispatch_queue_t lockQueue;
-@property NSMapTable<NSNumber *, NSObject *> *identifiersToInstances;
-@property NSMapTable<NSObject *, NSNumber *> *instancesToIdentifiers;
+@property NSMapTable<NSObject *, NSNumber *> *identifiers;
+@property NSMapTable<NSNumber *, NSObject *> *weakInstances;
+@property NSMapTable<NSNumber *, NSObject *> *strongInstances;
 @end
 
 @implementation FWFInstanceManager
-- (instancetype)init {
+- (instancetype)initWithDeallocCallback:(FWFOnDeallocCallback)callback {
+  self = [self init];
   if (self) {
+    _deallocCallback = callback;
     _lockQueue = dispatch_queue_create("FWFInstanceManager", DISPATCH_QUEUE_SERIAL);
-    _identifiersToInstances = [NSMapTable strongToStrongObjectsMapTable];
-    _instancesToIdentifiers = [NSMapTable strongToStrongObjectsMapTable];
+    _identifiers = [NSMapTable weakToStrongObjectsMapTable];
+    _weakInstances = [NSMapTable strongToWeakObjectsMapTable];
+    _strongInstances = [NSMapTable strongToStrongObjectsMapTable];
   }
   return self;
 }
 
-- (void)addInstance:(nonnull NSObject *)instance withIdentifier:(long)instanceIdentifier {
-  NSAssert(instance && instanceIdentifier >= 0,
-           @"Instance must be nonnull and identifier must be >= 0.");
+- (void)addFlutterCreatedInstance:(NSObject *)instance withIdentifier:(long)instanceIdentifier {
+  NSParameterAssert(instance);
+  NSParameterAssert(instanceIdentifier >= 0);
   dispatch_async(_lockQueue, ^{
-    [self.instancesToIdentifiers setObject:@(instanceIdentifier) forKey:instance];
-    [self.identifiersToInstances setObject:instance forKey:@(instanceIdentifier)];
+    [self addInstance:instance withIdentifier:instanceIdentifier];
+  });
+}
+
+- (long)addHostCreatedInstance:(nonnull NSObject *)instance {
+  NSParameterAssert(instance);
+  long identifier = -1;
+  dispatch_sync(_lockQueue, ^{
+    long identifier;
+    do {
+      // Identifiers are generated randomly to avoid collisions with objects
+      // created simultaneously by Flutter.
+      // Values are >= 2^16 and Flutter is expected to use values < 2^16.
+      identifier = arc4random_uniform(65536) + 65536;
```

Comment:
If you're on a queue, why not just use an incrementing counter (with wraparound) in the segmented space?